lib/client/keystore.go

```diff
@@ -608,52 +597,69 @@ func (fs *FSLocalKeyStore) GetKnownHostKeys(hostname string) ([]ssh.PublicKey, e
 	return retval, nil
 }
 
-// dirFor returns the path to the session keys for a given host. The value
-// for fs.KeyDir is typically "~/.tsh", sessionKeyDir is typically "keys",
-// and proxyHost typically has values like "proxy.example.com".
-//
-// If the create flag is true, the directory will be created if it does
-// not exist.
-func (fs *FSLocalKeyStore) dirFor(proxyHost string, create bool) (string, error) {
-	dirPath := filepath.Join(fs.KeyDir, sessionKeyDir, proxyHost)
-
-	if create {
-		if err := os.MkdirAll(dirPath, profileDirPerms); err != nil {
-			fs.log.Error(err)
-			return "", trace.ConvertSystemError(err)
+// AddKnownHostKeys adds a new entry to 'known_hosts' file
+func (fs *fsLocalNonSessionKeyStore) AddKnownHostKeys(hostname string, hostKeys []ssh.PublicKey) (retErr error) {
+	fp, err := os.OpenFile(filepath.Join(fs.KeyDir, fileNameKnownHosts), os.O_CREATE|os.O_RDWR, 0640)
+	if err != nil {
+		return trace.ConvertSystemError(err)
+	}
+	defer utils.StoreErrorOf(fp.Close, &retErr)
+	// read all existing entries into a map (this removes any pre-existing dupes)
+	entries := make(map[string]int)
+	output := make([]string, 0)
+	scanner := bufio.NewScanner(fp)
+	for scanner.Scan() {
+		line := scanner.Text()
+		if _, exists := entries[line]; !exists {
+			output = append(output, line)
+			entries[line] = 1
 		}
 	}
```

Comment:
I know it's existing code but this is missing a check for `scanner.Err()` as it might have exited due to an error.